chore: add callouts and troubleshooting guide for configuration (#6967)
aws-amplify/docs: src/pages/[platform]/build-a-backend/troubleshooting/library-not-configured/index.mdx

    'nextjs',
    'react',
    'react-native',
    'vue'
  ],
  canonicalObjects: [
    {
      platforms: [
        'angular',
        'react',
        'react-native',
        'vue',
        'nextjs',
        'javascript'
      ],
      canonicalPath: '/angular/build-a-backend/troubleshooting/migrate-from-javascript-v5-to-v6/'
    }
  ]
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

If you are running into the `Library Not Configured` error message and have called `Amplify.configure` in your project, your Amplify API is most likely being called before `Amplify.configure`. This can happen in a few different scenarios. Below are a list of things you can check to troubleshoot this issue.

## Call `Amplify.configure` in the root of your project

Make sure you are calling `Amplify.configure` in the root file of your project. The root file of your app may be different depending on your frontend framework:

* Vue: **src/main.ts** or **src/main.js**
* React: **src/main.tsx** or **src/main.jsx**
* Next:

> * Angular: **main.ts** or **main.js**

Author: Still need to update these paths and the verbiage to specify that these will be the default paths in common frameworks.

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Library Not Configured',

>   description: 'Addressing Library Not Configured error message',

Reviewer: ```suggestion
  description: 'Addressing Library Not Configured error messages',
```

</Callout>

## Check for module-level Amplify API invocations

When Amplify APIs are used outside of your application lifecycle, there is a risk that a javascript bundler may place that API call before `Amplify.configure`. Module-level function calls (calls at the top-level of a file), are generally evaluated in the order that they are imported.

Below is an example of code that will likely result in a `Library Not Configured` error:

```javascript title="index.ts"
import { Amplify } from 'aws-amplify';
import ComponentX from 'module-fetch-auth';

// fetchAuthSession() in ComponentX executed on import

Amplify.configure();

export default function App() {
  return (
    <div>
        <ComponentX />
    </div>
  );
}
```

```javascript title="module-fetch-auth.ts"
import { fetchAuthSession } from 'aws-amplify/auth';

fetchAuthSession(); // Will throw the Library Not Configured error

export default function ComponentX() {
  return (
    <div className="box">
      ...
    </div>
  );
}
```

> This can also happen when using Layouts (as in NextJS) and calling Amplify API’s in child components at the module-level (at the top of your file/module). See below for an example of this issue:

Reviewer: ```suggestion
This can also happen when using Next.js Layouts and calling Amplify API’s in child components at the module-level (at the top of your file/module). See below for an example of this issue:
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {

>   title: 'Library Not Configured',

Reviewer 2: ```suggestion
  title: 'Troubleshoot "Library Not Configured" errors',
```

Author: I updated this to 'Troubleshoot configuration errors' since LibaryNotConfigured is not the actual error message sent.

        'react-native',
        'vue',
        'nextjs',
        'javascript'
      ],
      canonicalPath: '/angular/build-a-backend/troubleshooting/migrate-from-javascript-v5-to-v6/'
    }
  ]
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

If you are running into a missing configuration or `NoCredentials` error message and have called `Amplify.configure` in your project, your Amplify API is most likely being called before `Amplify.configure`. This can happen in a few different ways. Below are three possibilities you can check to troubleshoot this issue.

## Call `Amplify.configure` in the root of your project

Make sure you are calling `Amplify.configure` in the root file of your project. The root file of your app may be different depending on your frontend framework. The current default for some common frameworks are listed below (if you are not using TypeScript the `ts` and `tsx` extensions would be `js` and `jsx`):

* Vue.js: **src/main.ts**
* React: **src/main.tsx**
* Angular: **src/main.ts**

> * Next.js Page Router: **pages/_app.tsx** or **src/pages/_app.tsx**

Reviewer 3: You could also add Nuxt to this list.  You should run the configuration in the `app.vue` [file](https://nuxt.com/docs/guide/directory-structure/app) in the root. Or in a plugins file, as [recommended here](https://docs.amplify.aws/react/build-a-backend/server-side-rendering/nuxt/#pageMain).

Author: Oh nice, thanks!

<Callout>

If you are using the Next.js App Router, you can follow the suggestions in our [Next.js documentation](/[platform]/build-a-backend/server-side-rendering/nextjs/#configure-amplify-library-for-client-side-usage) for root-level configuration. Keep in mind that if you are calling any APIs at the module-level (i.e. at the top of your file) in any of the Child components, you may still run into this issue.

</Callout>

## Check for module-level Amplify API invocations

> When Amplify APIs are used outside of your application lifecycle, there is a risk that a javascript bundler may place that API call before `Amplify.configure`. Module-level function calls (calls at the top-level of a file), are generally evaluated in the order that they are imported.

Reviewer 2: ```suggestion
When Amplify APIs are used outside of your application lifecycle, there is a risk that a JavaScript bundler may place that API call before `Amplify.configure`. Module-level function calls (calls at the top-level of a file), are generally evaluated in the order that they are imported.
```

Below is an example of code that will likely result in a missing configuration or `NoCredentials` error message:

> ```javascript title="index.ts"

Reviewer 4: nit: I think ```tsx is supported (I may be wrong) for better highlighting with jsx syntax

Author: Lovely

import { Amplify } from 'aws-amplify';
import ComponentX from 'module-fetch-auth';

// fetchAuthSession() in ComponentX executed on import

Amplify.configure();

export default function App() {
  return (
    <div>
        <ComponentX />
    </div>
  );
}
```

```tsx title="module-fetch-auth.ts"
import { fetchAuthSession } from 'aws-amplify/auth';

fetchAuthSession(); // Will throw "AuthUserPoolException: Auth UserPool not configured."

export default function ComponentX() {
  return (
    <div className="box">
      ...
    </div>
  );
}
```

> This can also happen when using Next.js Layouts and calling Amplify APIs in child components at the module-level (at the top of your file/module). See below for an example of this issue:

Reviewer 2: ```suggestion
This error can also happen when using Next.js Layouts and calling Amplify APIs in child components at the module-level (at the top of your file/module). See below for an example of this issue:
```

```javascript title="layout.tsx"
import ConfigureAmplifyClientSide from '@/ConfigureAmplifyClientSide';

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className="container py-6">
        <>
          <ConfigureAmplifyClientSide />
          {children}
        </>
      </body>
    </html>
  );
}
```

```javascript title="ConfigureAmplifyClientSide.tsx"
import { Amplify } from "aws-amplify";

Amplify.configure(config, { ssr: true });

export default function ConfigureAmplifyClientSide() {
  return null;
}
```

```javascript title="page.tsx"
import { fetchAuthSession } from "aws-amplify/auth";

// The layout calls configure, but fetchAuthSession ends up executing first
// Will throw "AuthUserPoolException: Auth UserPool not configured."
fetchAuthSession().then((session) => {
  console.log(session);
});

export default function HomePage() {
  return (
    <div className="box">
      ...
    </div>
  );
}
```

## Configure Amplify on each page of a multi-page app

> If you are working in a multi-page app, make sure `Amplify.configure` is called for each page.

Reviewer 4: What do you think of this alternative or similar:

If you are working in a multi-page app, you can call `Amplify.configure` in a common source file and then import it into each page.

Author: Ooo that is really nice
